infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LayoutAssetDao.kt

```diff
@@ -361,36 +307,59 @@ abstract class LayoutAssetDao<T : LayoutAsset<T>>(
     override fun deleteDrafts(branch: LayoutBranch): List<DaoResponse<T>> = deleteDraftsInternal(branch)
 
     private fun deleteDraftsInternal(branch: LayoutBranch, id: IntId<T>? = null): List<DaoResponse<T>> {
-        assertMainBranch(branch)
         val sql = """
             delete from ${table.fullName}
             where draft = true 
               and (:id::int is null or :id = id or :id = official_row_id)
+              and design_id is not distinct from :design_id
             returning 
-              coalesce(official_row_id, id) as official_id,
+              coalesce(official_row_id, design_row_id, id) as official_id,
               id as row_id,
               version as row_version
         """.trimIndent()
         jdbcTemplate.setUser()
-        return jdbcTemplate.query<DaoResponse<T>>(sql, mapOf("id" to id?.intValue)) { rs, _ ->
+        return jdbcTemplate
+            .query<DaoResponse<T>>(
+                sql,
+                mapOf("id" to id?.intValue, "design_id" to branch.designId?.intValue)
+            ) { rs, _ ->
             rs.getDaoResponse("official_id", "row_id", "row_version")
         }.also { deleted -> logger.daoAccess(DELETE, table.fullName, deleted) }
     }
 }
 
-private fun officialFetchSql(table: DbTable, fetchType: FetchType) = """
-    select o.id, o.version
-    from ${table.fullName} o left join ${table.fullName} d on d.official_row_id = o.id
-    where (o.id ${idOrIdsEqualSqlFragment(fetchType)} or d.id ${idOrIdsEqualSqlFragment(fetchType)}) and o.draft = false
-"""
-
-private fun draftFetchSql(table: DbTable, fetchType: FetchType) = """
-    select o.id, o.version 
-    from ${table.fullName} o
-    where o.official_row_id ${idOrIdsEqualSqlFragment(fetchType)} 
-       or (o.id ${idOrIdsEqualSqlFragment(fetchType)} 
-          and not exists(select 1 from ${table.fullName} d where d.official_row_id = o.id))
-"""
+private fun fetchContextVersionSql(table: LayoutAssetTable, fetchType: FetchType) =
```

Comment:
Pelottavan SQL:n ajatus on: Meille on annettu jonkin olion ID (tai joukko iideitä) _jossain_ kontekstissa, ja etsittävänä on tämän olion versio jossain layout-kontekstissa.

Asia menisi helpohkosti sillä, että haetaan vaan olion official_id, ja sitten tuolta in_layout_contextista haetaan samalla official_id:llä rivi. Mutta tämä ei käy, koska funktion sisällä official_id on määritelty useamman kentän coalescena, ja postgres ei juuri osaa optimoida hakua moisen läpi, vaan tekee siitä vaan seq scannin, joka katsoo jokaisesta kontekstin oliosta yksi kerrallaan, onko nyt tämä se, jonka official_id on sama kuin annettu.

Siksi mennään tällä monimutkaisemmalla haulla: lookup-relaatioon haetaan sekä ylös- että alaspäin perintähierarkiasta kaikki samaa oliota vastaavat rivit, ja sitten suodatetaan joinaamalla in_layout_contextista juuri se rivi, joka on oikeassa kontekstissa. Luultavasti tässä haussa on jonkin verran päällekkäisyyttä ja teoreettista optimointivaraa, mutta perustasolla se tuntuu toimivan riittävän kohtuuhyvällä perffillä, ja hakusuunnitelmissa ei ole ylimääräisiä seq scanneja.